src/webui/webapplication.cpp

```diff
@@ -914,76 +800,106 @@ void WebApplication::action_command_addCategory()
         return;
     }
 
-    if (!isPublicScope() && !sessionActive()) {
-        status(403, "Forbidden");
-        return;
-    }
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isPublicAPI(methodName) && !sessionActive()) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (actions_.value(scope_).value(action_) != 0) {
-        (this->*(actions_[scope_][action_]))();
+        if (!QMetaObject::invokeMethod(this, methodName.toLatin1().constData())) {
+            status(404, "Not Found");
+            qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        }
     }
     else {
-        status(404, "Not Found");
-        qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        getWebUIFile();
+    }
+}
+
+void WebApplication::configure()
+{
+    const auto pref = Preferences::instance();
+    m_username = pref->getWebUiUsername();
+    m_password = pref->getWebUiPassword();
+
+    const QString rootFolder = Utils::Fs::expandPathAbs(
+                !pref->isAltWebUiEnabled() ? WWW_FOLDER : pref->getWebUiRootFolder());
+    if (rootFolder != m_rootFolder) {
+        m_translatedFiles.clear();
+        m_rootFolder = rootFolder;
     }
 }
 
-void WebApplication::parsePath()
+void WebApplication::printFile(const QString &path)
 {
-    if (request().path == "/") action_ = WEBUI_ACTION;
+    const QString type {getContentType(path)};
+    const bool isTranslatable {(type == Http::CONTENT_TYPE_HTML) || (type == Http::CONTENT_TYPE_JS)};
+    const QDateTime lastModified {QFileInfo(path).lastModified()};
 
-    // check action for requested path
-    QStringList pathItems = request().path.split('/', QString::SkipEmptyParts);
-    if (!pathItems.empty() && actions_.contains(pathItems.front())) {
-        scope_ = pathItems.front();
-        pathItems.pop_front();
+    // find translated file in cache
+    if (isTranslatable) {
+        auto it = m_translatedFiles.constFind(path);
+        if ((it != m_translatedFiles.constEnd()) && (lastModified <= (*it).lastModified)) {
+            print((*it).data, type);
+            return;
+        }
     }
 
-    if (!pathItems.empty() && actions_[scope_].contains(pathItems.front())) {
-        action_ = pathItems.front();
-        pathItems.pop_front();
+    QFile file {path};
+    if (!file.open(QIODevice::ReadOnly)) {
+        qDebug("File %s was not found!", qUtf8Printable(path));
+        status(404, "Not Found");
+        return;
+    }
+    
+    if (file.size() > maxAllowedFilesize()) {
+        qWarning("%s: exceeded the maximum allowed file size!", qUtf8Printable(path));
+        status(500, "Internal Server Error");
+        print(QObject::tr("Exceeded the maximum allowed file size!"), Http::CONTENT_TYPE_TXT);
+        return;
     }
 
-    args_ = pathItems;
-}
+    QByteArray data {file.readAll()};
```

Comment:
For strictness I would use `file.read(maxAllowedFilesize())`.

Reply:
Is it possible to read more than size() bytes? Can the file be changed after we've opened it?

Reply:
>Can the file be changed after we've opened it?

Good point, I missed it.